### ISSUE TITLE
test for memory leaks (brand=phet)

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/graphing-quadratics#49]
Do memory profiling and test for leaks, similar to what was done in https://github.com/phetsims/equality-explorer/issues/64.  

Use a built version.

**Comment by pixelzoom** [comment]
Results of testing using:
* [1.0.0-dev.30](https://phet-dev.colorado.edu/html/graphing-quadratics/1.0.0-dev.30/phet/graphing-quadratics_en_phet.html) with `?fuzzMouse`
* Chrome 69.0.3497.100
* macOS 10.11.6

| Minutes | Heap Size (MB) |
| --- | --- |
| 0 | 39.9 |
| 5 | 71.7 |
| 10 | 88.7 |
| 15 | 104 |
| 20 | 113 |
| 25 | 124 |
| 30 | 137 |
| 40 | 151 |

**Comment by pixelzoom** [comment]
Running in requirejs mode, I see similar continuous growth of the heap.  Looking at heap summary, `QuadraticNode` consistently has '0' in the "# Deleted" column, so I'll start looking there.

**Comment by pixelzoom** [comment]
`QuadraticNode` did indeed have a memory leak, plugging the in above commit.

I inspected all calls to `link`, `addInputListener`, etc. again, and I don't see anything else involved that is dynamically created.  So I'll proceed with another round of heap testing.

**Comment by pixelzoom** [comment]
This suggests that a slow leak is acceptable, or perhaps even "normal".

Slack dev-public:

> Sam Reid [9:56 AM]
Memory leak testing Graphing Quadratics PhET-iO over the last 5 minutes doesn’t suggest any serious memory leaks (it’s just the normal slow rising pattern).  I also don’t see any tandems being added after the sim starts up, so that’s a good indicator that there are no phet-io related memory leaks.
>
> Chris Malley [9:59 AM]
Thanks, that matches what I see.  I wasn’t aware of this “normal slow rising pattern”.  When I tested Equality Explorer (5 screens), it started at 71MB and then oscillated around 82MB.

**Comment by pixelzoom** [comment]
Results of testing for a longer period, using

* requirejs mode with master (sha 5c810e9debe45272ee9cc0c7ca1ea661ed21ca83)
* `?fuzzMouse`
* Chrome 69.0.3497.100
* macOS 10.11.6

| Hours | Heap Size (MB) |
| --- | --- |
| 0 | 42.0 |
| 1 | 53.3 |
| 2 | 57.0 |
| 3 | 61.1 |
| 4 | 64.2 |

Note that time is in hours. After the first hour, we're growing by ~3.5MB per hour. 

@ariel-phet Your call on whether this amount of leakage is acceptable. If not, I suspect that it's typical of all sims, and I'm going to need some assistance tracking it down.

**Comment by pixelzoom** [comment]
@ariel-phet and I talked about this via Zoom.   This leak is slow enough so that it shouldn't be a problem in practice, and does not prevent us from publishing 1.0.   It would be nice to understand where the leak is coming from, and whether it's impacting other sims, so I'll ask other developers about known leaks.  Then I'll close this issue.

**Comment by pixelzoom** [comment]
There is apparently a memory leak in scenery that could be responsible for the leak observed here. 

Slack conversation with @chrisklus and @jonathanolson:

Chris Malley [2:25 PM]
Graphing Quadratics memory leak testing is revealing a slow memory leak.  While chatting with Ariel about this, he said to ping you guys about a potential scenery leak that was recently discovered during work on sun buttons.  Does that ring any bells?

Chris Klusendorf [2:26 PM]
Yeah, if the memory profile shows Colors or Emitters leaking, that’s related to the scenery issue

Chris Malley [2:27 PM]
I can’t tell exactly what’s leaking. Under what conditions do Colors and Emitters leak?
Or is there a GitHub issue that describes this?

Chris Klusendorf [2:29 PM]
In snapshot comparisons, the deltas should give some information, right? The leak is from Colors uses in gradients, like buttons
There should be a github issue but
I don’t think there is one yet, we had just finished finding the source of the problem yesterday

Chris Malley [4:20 PM]
One of the things that shows up as in heap comparison as “# Deleted = 0” is SVGGradientStop.  Related? (edited)
Ah, my bad… “# Delta = 0”, so looks like it’s cleaned up.
Comparing 2 heap snapshots taken 20 minutes apart, Emitter and Color are both towards the top of of the list based on #Delta and Alloc. Size.

Does this look indicative of the Color/Emitter leak?

![screenshot_114](https://user-images.githubusercontent.com/3046552/46820813-50b04680-cd44-11e8-98ec-5f981682d4f0.jpg)

Jonathan Olson [4:35 PM]
Potentially looks like it yes

Chris Malley [4:37 PM]
Graphing Quadratics creates all of its buttons at startup, and does not explicitly create any Colors or Emitters.  Can you summarize the problem?

Jonathan Olson [4:39 PM]
Essentially an internal array keeps growing when a specific case of a gradient with two color stops using the same color Property (when the value changes AND it is displayed, and then later not displayed)

Chris Malley [4:40 PM]
I’m still trying to parse that :)
Is that something that can happen with a plain old sun pseudo-3D button that exists for the lifetime of the sim?

Jonathan Olson [4:43 PM]
It presumably can as long as it goes back and forth between having drawables and not (like getting removed, maybe screen switches) and one of the gradients changes colors dynamically

Chris Malley [4:44 PM]
No buttons are removed from the scenegraph.  No colors are changed explicitly.
But I don’t know what sun buttons are doing internally with colors and gradients.

Chris Klusendorf [4:53 PM]
I’m not sure how the buttons work either but could something like clicking reset all or the visibility button cause it? Those both have gradients and are changing color on press

Chris Malley [4:53 PM]
that’s what I was wondering.

**Comment by pixelzoom** [comment]
I don't understand the nature of the scenery leak and whether it is relevant to this sim. Labeling with meeting:developer to see if anyone can clarify for me.

**Comment by chrisklus** [comment]
@pixelzoom https://github.com/phetsims/scenery/issues/879

**Comment by pixelzoom** [comment]
phetsims/scenery#879 is "PaintObserver internal failure and memory leak".  And  I do see PaintObserver with "#Deleted = 0" in Chrome heap comparisons.

**Comment by jessegreenberg** [comment]
@ariel-phet said that it is good to understand where the leak is coming from but that this doesn't need to block publication.

**Comment by pixelzoom** [comment]
10/11/18 dev meeting notes:
- the current rate of leaking is acceptable for 1.0 publication 
- phetsims/scenery#879 does not need to be addressed for 1.0 publication
- once phetsims/scenery#879 is fixed, it will be tested in Graphing Quadratics (noted in that issue)
- going forward, memory leak testing should be done with `?fuzzMouse&fuzzTouch`
- OK to close this issue

**Comment by pixelzoom** [comment]
phetsims/scenery#879 leak is fixed.  Reopening to document whether it fixes the small leak that was observed in this sim.

**Comment by pixelzoom** [comment]
Results of testing after phetsims/scenery#879 leak was fixed.

* requirejs mode with master (sha 903073f4748a45a9fcbe44ec0a651c0d5f10db98)
* `?fuzz`
* Chrome 69.0.3497.100
* macOS 10.11.6

| Minutes | Heap Size (MB) |
| --- | --- |
| 0 | 43.4 |
| 5 | 49.2 |
| 10 | 49.8 |
| 15 | 51.1 |
| 20 | 51.6 |
| 25 | 52.2 |
| 30 | 52.5 |

51.4 MB immediately after pressing the "Collect garbage" button.

Not too different from the results in https://github.com/phetsims/graphing-quadratics/issues/49#issuecomment-428441974, looks like there's still a slow continual increase.  Nothing obvious in heap comparisons.

**Comment by pixelzoom** [comment]
See @jonathanolson's testing in https://github.com/phetsims/scenery/issues/879#issuecomment-430099577.  I'm convinced that the main leak (PaintObserver) is fixed, and we're only leaking 250kB/hour with fuzz testing.   So I'm happy to close this issue again.
